src/main/java/com/metamx/emitter/core/HttpPostEmitter.java

```diff
@@ -464,22 +474,23 @@ private void emitLargeEvent(byte[] eventBytes)
       System.arraycopy(eventBytes, 0, buffer, bufferOffset, eventBytes.length);
       bufferOffset += eventBytes.length;
       bufferOffset = batchingStrategy.writeBatchEnd(buffer, bufferOffset);
-      if (sendWithRetries(buffer, bufferOffset, 1)) {
+      if (sendWithRetries(buffer, bufferOffset, 1, true)) {
         buffersToReuse.add(buffer);
         approximateBuffersToReuseCount.incrementAndGet();
       } else {
-        failedBuffers.add(new FailedBuffer(buffer, bufferOffset, 1));
+        limitFailedBuffersSize();
+        failedBuffers.addLast(new FailedBuffer(buffer, bufferOffset, 1));
         approximateFailedBuffersCount.incrementAndGet();
       }
     }
 
     private void tryEmitOneFailedBuffer()
     {
-      FailedBuffer failedBuffer = failedBuffers.peek();
+      FailedBuffer failedBuffer = failedBuffers.peekFirst();
       if (failedBuffer != null) {
-        if (sendWithRetries(failedBuffer.buffer, failedBuffer.length, failedBuffer.eventCount)) {
+        if (sendWithRetries(failedBuffer.buffer, failedBuffer.length, failedBuffer.eventCount, false)) {
```

Comment:
Each individual retry is still timing out, so it's limited to approximately 3 timouts (because we do 3 retires), so buffers to emit could get ~ 3 more batches at this time